### ISSUE TITLE
Failing to resolve and publish `vscjava.vscode-gradle`

### ISSUE THREAD

**Issue by vince-fugnitto** [open-vsx/publish-extensions#770]
### Description

There is an issue when attempting to resolve the [Gradle for Java](https://open-vsx.org/extension/vscjava/vscode-gradle) causing the extension to never be published at a newer version. This means that only a single version exists on Open-VSX (~5 month old pre-release version).

The extension is referenced under `extensions.json` but there are issues resolving the extension during [CI](https://github.com/open-vsx/publish-extensions/actions/runs/10087619205/job/27892056317#step:12:7145):

https://github.com/open-vsx/publish-extensions/blob/6c39f3f4bd307cc411b614d58eacb7b556db2460/extensions.json#L1425-L1427

```
[FAIL] Could not process extension: {
  "extension": {
    "id": "vscjava.vscode-gradle",
    "repository": "https://github.com/microsoft/vscode-gradle"
  },
  "context": {
    "msVersion": "3.15.0",
    "msLastUpdated": "2024-07-19T03:23:19.677Z",
    "msInstalls": 3052039,
    "msPublisher": "vscjava",
    "ovsxVersion": "3.13.2024011802",
    "ovsxLastUpdated": "2024-02-13T19:21:03.109Z"
  }
}
vscjava.vscode-gradle: failed to resolve
```

### Reaspm

The extension cannot be resolved due to a couple of reasons:
- the extension's sources (ex: `package.json`) can be found under the `/extensions` folder https://github.com/microsoft/vscode-gradle/tree/develop/extension not the root of the repository like the `extensions.json` references https://github.com/microsoft/vscode-gradle 
- the build and packaging of the extension is more complex and cannot be automatically inferred by the CI (proper build commands need to be used)

### Additional Information

- https://github.com/open-vsx/publish-extensions/issues/730

cc @kineticsquid

**Comment by kineticsquid** [comment]
We had issues with this extension before, don't recall the detail, which is why I published the one version. I've just updated it with the latest. Though, this won't resolve the auto publishing.

**Comment by Kenneth-Sills** [comment]
Yup, that seems to be the issue, @vince-fugnitto. Here's a rough two-step process for getting this working again.

First, the [`extensions.json` entry](https://github.com/EclipseFdn/publish-extensions/blob/3b3ac71b97ce6e5d25aeea7eb066ff61c6e00dbc/extensions.json#L1438) will need to have a [`"location": "extension"`](https://github.com/EclipseFdn/publish-extensions/blob/3b3ac71b97ce6e5d25aeea7eb066ff61c6e00dbc/extensions-schema.json#L15) added. That should at least get us past the `failed to resolve` issue but builds will start failing.

Then, we need to add a [`custom`](https://github.com/EclipseFdn/publish-extensions/blob/3b3ac71b97ce6e5d25aeea7eb066ff61c6e00dbc/extensions-schema.json#L27) to the entry, because the Gradle plugin needs to build out both a separately-distributed build server repository and then the repo-included language server using Gradle before bundling the extension VSIX. Build instructions are documented [here](https://github.com/microsoft/vscode-gradle/blob/develop/CONTRIBUTING.md#running-the-project):

```sh
cd extension
git clone https://github.com/microsoft/build-server-for-gradle.git
../gradlew buildJars
cd ..
./gradlew build
cd extension
npx vsce package -o extension.vsix
```

That top-level Gradle build [handles NPM installation & builds](https://github.com/microsoft/vscode-gradle/blob/4794597153e32c4a6c09f1c12801a10e6ebd20f0/extension/build.gradle#L252).

> [!note]
> 
> From what I can see, [the workflow for building extensions](https://github.com/EclipseFdn/publish-extensions/blob/3b3ac71b97ce6e5d25aeea7eb066ff61c6e00dbc/.github/workflows/publish-extension.yml#L53) is a touch brittle in how it sets up specifically Node 20[^1]  and Java 21[^2] - without NVM/Volta or SDKMan to use extension-specific versions - but that should *probably* be fine since upstream requests 18 & 21, respectively.

[^1]: Via https://github.com/actions/setup-node
[^2]: Via https://github.com/actions/setup-java

None of this is tested, just ripped from their docs and looking at this project's source. So any assignee will need to test it actually works. 😅 

I've also [made a ticket upstream](https://github.com/microsoft/vscode-gradle/issues/1696) to request a new `3.17.0` release tag so we get the most recent version.

**Comment by kineticsquid** [comment]
@vince-fugnitto @Kenneth-Sills Do you need me to do anything?
